apps/src/code-studio/header.js

```diff
@@ -173,21 +189,63 @@ function shareProject() {
 
     ReactDOM.render(
       <Provider store={getStore()}>
-        <ShareDialog
+        <ConnectedShareDialog
           i18n={i18n}
           icon={appOptions.skin.staticAvatar}
           shareUrl={shareUrl}
           isAbusive={dashboard.project.exceedsAbuseThreshold()}
+          isOwner={dashboard.project.isOwner()}
+          isPublished={dashboard.project.isPublished()}
           channelId={dashboard.project.getCurrentId()}
           appType={appType}
           onClickPopup={popupWindow}
           // TODO: Can I not proliferate the use of global references to Applab somehow?
           onClickExport={window.Applab ? window.Applab.exportApp : null}
+          onShowPublishDialog={showPublishDialog}
+          onUnpublish={unpublishProject}
           canShareSocial={canShareSocial}
         />
       </Provider>,
       dialogDom
     );
+
+    getStore().dispatch(showShareDialog());
+  });
+}
+
+function showPublishDialog() {
+  var publishDialog = document.getElementById('publish-dialog');
+  if (!publishDialog) {
+    publishDialog = document.createElement('div');
+    publishDialog.setAttribute('id', 'publish-dialog');
+    document.body.appendChild(publishDialog);
+  }
+
+  ReactDOM.render(
+    <Provider store={getStore()}>
+      <PublishDialog
+        onConfirmPublish={publishProject}
+      />
+    </Provider>,
+    publishDialog
+  );
+
+  const projectId = window.dashboard.project.getCurrentId();
+  const projectType = dashboard.project.getStandaloneApp();
+  getStore().dispatch(hideShareDialog());
+  getStore().dispatch(showPublishDialogAction(projectId, projectType));
+}
+
+function publishProject(projectId, projectType) {
+  getStore().dispatch(publishProjectAction(projectId, projectType))
+    .then(projectData => {
+      window.dashboard.project.setPublishedAt(projectData.publishedAt);
+    });
+}
+
+function unpublishProject(projectId) {
+  getStore().dispatch(unpublishProjectAction(projectId)).then(() => {
+    window.dashboard.project.setPublishedAt(null);
```

Comment:
I wonder if a cleaner approach here would be instead of passing in callbacks to our component, the component just dispatches actions, and then here we have a listener that calls our dashboard APIs, i.e.
```
let state = {};
store.subscribe(() => {
  lastState = state;
  state = store.getState();
  if (state.lastPublishedAt !== lastState.lastPublishedAt) {
    setPublishedAt(state.lastPublishedAt);
  }
});
```
This way our UI is a little more declarative (we have our React components updating to match redux store via connect, and we have our non-React UI updating via this listener).